Refactor cron event to validate individual URLs

dhaval-parekh opened the pull request:
## Summary

* Update `URLValidationCron` to implement a queue for validating URLs on an hourly basis. Once a week the URLs returned by `ScannableURLProvider` are queued up, and then on an hourly basis each URL is validated one-by-one until there are none left. If the validated environment changes at any point (e.g. changing the template mode) then the queue is cleared and new URLs are queued up for validation.
* Update `RecurringBackgroundTask` to support re-scheduling when the interval changes.
* Allow passing `$options` into `AMP_Theme_Support::get_supportable_templates()` to obtain results for options which are not saved in the DB.
* Add more options to the validated environment, including what templates are suppoerted and what Reader theme is selected.
* Deprecate the obsolete `--force` flag on `wp amp validation run` which no longer has any effect.
* Refactor `ScannableURLProvider` to eliminate the need for `URLScanningContext`. Ensure that only URLs for supported templates (and AMP-enabled posts) are returned by `get_urls`.
* Eliminate the unused `amp_url_validation_sleep_time` filter.
* Use the year archive URL as returned by `get_year_link()` rather than manually constructing it with query parameters. Ensure it will have posts on it.

Fixes #5750 

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

dhaval-parekh committed:
Update URL validation cron jobs

dhaval-parekh committed:
Add unit test cases

westonruter commented on src/BackgroundTask/CronBasedBackgroundTask.php:
```diff
@@ -72,4 +74,23 @@ abstract protected function get_event_name();
+		$cron_schedules = ( ! empty( $cron_schedules ) && is_array( $cron_schedules ) ) ? $cron_schedules : [];
+
+		$cron_schedules[ self::DEFAULT_INTERVAL_EVERY_TEN_MINUTES ] = [
+			'interval' => ( MINUTE_IN_SECONDS * 10 ),
```
```suggestion
			'interval' => ( 10 * MINUTE_IN_SECONDS ),
```

westonruter commented on src/BackgroundTask/CronBasedBackgroundTask.php:
```diff
@@ -72,4 +74,23 @@ abstract protected function get_event_name();
+
+		$cron_schedules[ self::DEFAULT_INTERVAL_EVERY_TEN_MINUTES ] = [
+			'interval' => ( MINUTE_IN_SECONDS * 10 ),
+			'display'  => esc_html__( 'Every 10 minutes', 'amp' ),
```
It's not known where this will be rendered, so escaping here is premature:

```suggestion
			'display'  => __( 'Every 10 minutes', 'amp' ),
```

westonruter commented on src/BackgroundTask/CronBasedBackgroundTask.php:
```diff
@@ -72,4 +74,23 @@ abstract protected function get_event_name();
+	 */
+	public function add_cron_schedules( $cron_schedules ) {
+
+		$cron_schedules = ( ! empty( $cron_schedules ) && is_array( $cron_schedules ) ) ? $cron_schedules : [];
```
Is there a known case where `$cron_schedules` is not an array?

```suggestion
		$cron_schedules = is_array( $cron_schedules ) ? $cron_schedules : [];
```

westonruter commented on src/Validation/URLValidationQueueCron.php:
```diff
@@ -0,0 +1,95 @@
+/**
+ * URLValidationCron class.
+ *
+ * @since 2.1
```
```suggestion
 * @since 2.2
```

westonruter commented on src/Validation/URLValidationQueueCron.php:
```diff
@@ -0,0 +1,95 @@
+			if ( empty( $url['url'] ) || empty( $url['type'] ) ) {
+				continue;
+			}
+
```
This condition probably isn't needed because `get_urls` _always_ returns an array with this shape. If it doesn't, then this should be considered an error and we'd want the PHP warning to be emitted.

```suggestion

			if ( empty( $url['url'] ) || empty( $url['type'] ) ) {
				continue;
			}

```

westonruter commented on src/Validation/URLValidationQueueCron.php:
```diff
@@ -0,0 +1,95 @@
+				continue;
+			}
+
+			$url_hash                      = md5( trim( $url['url'] ) );
```
Why trim?

```suggestion
			$url_hash                      = md5( $url['url'] );
```

westonruter commented on src/Validation/URLValidationQueueCron.php:
```diff
@@ -0,0 +1,95 @@
+	 */
+	public function process( ...$args ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		$urls                 = $this->scannable_url_provider->get_urls();
+		$validation_queue_key = 'amp_url_validation_queue';
```
This should probably be a constant which is referenced here and in `URLValidationCron`

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -96,15 +53,34 @@ public function __construct( BackgroundTaskDeactivator $background_task_deactiva
+			}
+
+			unset( $validation_queue[ $hash ] );
+			$count ++;
```
```suggestion
			$count++;
```

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -96,15 +53,34 @@ public function __construct( BackgroundTaskDeactivator $background_task_deactiva
 			}
 		}
+
+		update_option( $validation_queue_key, $validation_queue );
```
Make sure this option gets added to `uninstall.php`.

westonruter commented on src/Validation/URLValidationQueueCron.php:
```diff
@@ -0,0 +1,95 @@
+ * WP cron process to validate URLs in the background.
+ *
+ * @package AMP
+ * @since   2.1
```
```suggestion
 * @since   2.2
```

westonruter commented on src/Validation/URLValidationQueueCron.php:
```diff
@@ -0,0 +1,95 @@
+			}
+
+			$url_hash                      = md5( trim( $url['url'] ) );
+			$validation_queue[ $url_hash ] = $url;
```
Any reason in particular to use the hash? This could also be:

```suggestion
			if ( ! in_array( $url, $validation_queue, true ) ) {
				$validation_queue[] = $url;
			}
```

westonruter commented on src/Validation/URLValidationQueueCron.php:
```diff
@@ -0,0 +1,95 @@
+	 * @return string An existing interval name.
+	 */
+	protected function get_interval() {
+		return self::DEFAULT_INTERVAL_DAILY;
```
Let's update this to run once a week instead.

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -96,15 +53,34 @@ public function __construct( BackgroundTaskDeactivator $background_task_deactiva
+			return [];
+		}
+
+		$limit = 5;
+		$count = 1;
+
+		foreach ( $validation_queue as $hash => $url ) {
+			if ( empty( $url['url'] ) || empty( $url['type'] ) ) {
+				continue;
+			}
+
+			$response = $this->url_validation_provider->get_url_validation( $url['url'], $url['type'] );
+			if ( empty( $response ) || is_wp_error( $response ) ) {
+				continue;
+			}
+
+			unset( $validation_queue[ $hash ] );
+			$count ++;
+			if ( $limit < $count ) {
+				break;
 			}
 		}
```
Instead of looping over 5, I think this should rather just do one. If using the `in_array()` check suggested below, then this could be just:

```suggestion
		if ( empty( $validation_queue ) ) {
			return;
		}
		
		$entry = array_shift( $validation_queue );
		if ( isset( $entry['url'], $entry['type'] ) ) {
			$this->url_validation_provider->get_url_validation( $entry['url'], $entry['type'] );
		}
```

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -126,22 +102,6 @@ protected function get_event_name() {
-		 * @param int The number of seconds. Default 1. Setting to 0 or a negative numbers disables all throttling.
-		 */
-		return max( (int) apply_filters( 'amp_url_validation_sleep_time', self::DEFAULT_SLEEP_TIME ), 0 );
+		return self::DEFAULT_INTERVAL_EVERY_TEN_MINUTES;
```
I think this could be changed to run hourly, to reuse existing schedule and since it doesn't seem necessary to run quickly.

dhaval-parekh committed:
Validate single URL on URL validation cron, Update test cases

dhaval-parekh committed:
Hard core the site option key that need to delete

dhaval-parekh committed:
Merge branch 'develop' into dev-tool/5750-refactor-url-validation-cron

dhaval-parekh committed:
Update comments

dhaval-parekh committed:
Remove unused references

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into dev-tool/5750-refactor-url-validation-cron

* 'develop' of github.com:ampproject/amp-wp: (28 commits)
  Use assertContains for array
  Use DependencyInjectedTestCase
  Add test assertions
  Update comments
  Update unit test cases
  Remove unwanted style, Fix notice style in mobile
  Revert some changes and update unit test case
  Remove changes from .github
  Remove template change notice from other pages
  Remove template change notice from other pages
  Fix unefined index issue in test case
  Fix unit test cases
  Fix unit test cases
  Fix unit test cases
  Fix unit test cases
  Cache result of page cache check and use that when needed
  Fix unit test cases
  Update test cases
  Remove unwanted localize variable
  Only show notice for flush page cache if page cache is enable
  ...

westonruter committed:
Remove return of empty array

westonruter committed:
Replace extraneous URLValidationProvider with direct AMP_Validation_Manager call

westonruter committed:
Update tests to use DependencyInjectedTestCase

westonruter commented on tests/php/src/Validation/URLValidationCronTest.php:
```diff
@@ -124,7 +134,7 @@ public function test_get_event_name() {
 	public function test_get_interval() {
 		$this->assertEquals(
-			URLValidationCron::DEFAULT_INTERVAL_DAILY,
+			URLValidationCron::DEFAULT_INTERVAL_HOURLY,
```
I just noticed this change is problematic because there is an existing daily schedule, and because of that the change to hourly is not resulting in the re-scheduling. It seems we need to do a call to `wp_get_scheduled_event()` instead of using `wp_next_scheduled()` in `RecurringBackgroundTask::schedule_event()`, and then re-schedule it if the `schedule` does not match.

westonruter replied:
Took my first stab at addressing this in 422d0f2cc224d848eb014a4a390d242f0594d2fc. I'm not sure it's best, however. Namely, if there was a daily schedule that is going to happen in 23 hours, but then the schedule gets changed to hourly, then the initial timestamp remains 23 hours in the future. I guess it would be better to unschedule it enturely.

westonruter replied:
There, I think this is better: bf25b0cb4999c2c4e9d70a1923082cec2dab0aaa.

westonruter committed:
Simplify URL validation scheduling by putting the queue in URLValidationCron

westonruter committed:
Re-schedule with new recurrence when schedule changes

westonruter committed:
Unschedule event with wrong recurrence instead of rescheduling it

dhaval-parekh committed:
Fix unit test case

dhaval-parekh committed:
Revert "Fix unit test case"

This reverts commit e7f51ea22c2ce39f6d61fccd374e1bb336093655.

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -96,15 +61,49 @@ public function __construct( BackgroundTaskDeactivator $background_task_deactiva
+				return null;
 			}
+
+			// @todo The URLs should be contextual based on the selected template mode, in particular only singular URLs should be included if using legacy Reader mode.
```
Also relates to site scanning in general in #6610. Only the available templates should be scanned.

westonruter replied:
Done in 5bbadeaa6d5311090e3b61e0a5fc2f167bf0d760

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -96,15 +61,49 @@ public function __construct( BackgroundTaskDeactivator $background_task_deactiva
+			$data
+		);
+
+		// If there are no URLs queued, then obtain a new set.
+		if ( empty( $data['urls'] ) ) {
```
A change in the validated environment (and enabled post types and templates) should also trigger this, since we need to obtain a new set of URLs.

westonruter replied:
Done in 55c0ea7ef0d9125e22b88099eb05fafe5b8be75c and ab2bff10f5dbe12766a36b71db045103de507f87

westonruter committed:
Ensure that AMP-enabled posts are scanned

westonruter committed:
Remove obsolete force argument for amp validation CLI command

westonruter commented on src/BackgroundTask/RecurringBackgroundTask.php:
```diff
@@ -37,13 +37,19 @@ final public function schedule_event( ...$args ) {
 
-		if ( $timestamp ) {
-			return;
+		$scheduled_event = wp_get_scheduled_event( $event_name, $args );
```
Ah, this function was introduced in WP 5.1.

westonruter committed:
Provide wp_get_scheduled_event() implementation for WP<5.1 and add tests

westonruter committed:
Try not using past time for scheduling event

westonruter committed:
Use daily schedule to test since available in older WP versions

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into dev-tool/5750-refactor-url-validation-cron

* 'develop' of github.com:ampproject/amp-wp: (92 commits)
  Fix grammar typos in comment
  Update Gutenberg package dependencies
  Bump eslint-plugin-jsdoc from 36.1.0 to 36.1.1
  Bump lint-staged from 11.1.2 to 11.2.3
  Bump @babel/core from 7.15.5 to 7.15.8
  Bump css-minimizer-webpack-plugin from 3.0.2 to 3.1.1
  Bump postcss from 8.3.8 to 8.3.9
  Add effective sandboxing level to meta generator and admin bar
  Make ValidationExemption more DRY and performant
  Opt to disable stylesheet processing when in level 1
  Add yet more covers tags for private methods in style sanitizer
  Add test coverage for AMP_Validation_Error_Taxonomy::get_error_title_from_code()
  Add additional covers tags for private methods in audio, iframe, and video sanitizers
  Remove empty style attributes add style sanitizer test coverage
  Work around coverage decrease by adding covers tags for private methods in img sanitizer
  Add test coverage for invalid at-rule inside keyframes
  Add test coverage for allowlisted properties in CSS declaration block
  Improve test coverage for ValidationExemption
  Bump eslint-plugin-react from 7.26.0 to 7.26.1
  Fix sanitization of keyframes outside of style[amp-keyframes]
  ...

westonruter committed:
Fix covers phpdoc tags

westonruter committed:
Use assertLessThanOrEqual rather than assertGreaterThanOrEqual

westonruter committed:
Fix covers phpdoc tag for wrap_block_callbacks

westonruter committed:
Account for wp_schedule_event() returning bool as of WP 5.1

westonruter committed:
Remove obsolete arg from URLScanningContext constructor

westonruter committed:
Remove unused amp_url_validation_limit_per_type filter

westonruter committed:
Allow options to be overridden when calling get_supportable_templates

westonruter committed:
Refactor ScannableURLProvider and URLValidationProvider into services

* Eliminate extraneous URLScanningContext.
* Use template conditionals as type labels in WP-CLI output.
* Prepare ScannableURLProvider to return URLs that don't correspond to the currently-set template mode for Site Scan in Onboarding Wizard.

westonruter committed:
Include more AMP settings in validated environment

westonruter committed:
Fix tests after 5bbadea

westonruter committed:
Clear out cron validation queue whenever validated environment changes

westonruter committed:
Remove obsolete test for amp_should_use_new_onboarding

westonruter commented:
@pierlon @dhaval-parekh Please review my changes.

westonruter committed:
Fix test assertions by using polyfilled TestCase

westonruter committed:
Remove unused DEFAULT_INTERVAL_WEEKLY and fix test_process_and_dequeue assertions

westonruter committed:
Use get_year_link() to obtain the year archive link and ensure it has posts on it

westonruter committed:
Ensure author links are not 404

westonruter committed:
Improve test coverage

westonruter committed:
Fix test for get_author_page_urls

westonruter committed:
Fix get_date_page test in WP<=5.1

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into dev-tool/5750-refactor-url-validation-cron

* 'develop' of github.com:ampproject/amp-wp: (25 commits)
  Skip test_get_theme_info when no active theme (in WP 4.9)
  Fix logic in get_theme_info to return parent and child theme both
  Use wp_get_theme() to obtain theme version on demand; account for sources with deleted themes/plugins
  Only obtain theme versions for parent theme and child theme
  Omit parent_theme field if child theme not active
  Use closure to call normalize_plugin_info
  Use class name constant
  Add since tag and sort imports
  Use imported WP_Query
  Ensure data is array before sending it through array_map
  Update message in AMP Support page
  Apply suggestions from PR
  Use class injector instead of fetching from Service class
  Add delayed in SupportLink class and update the unit test cases
  Add more unit test cases for SupportData()
  Apply suggestions from PR and update unit test cases
  Update unit test case
  Fix unit test cases
  Fix fatal error in WP 4.9
  Update logic to get error log
  ...

westonruter commented on src/Validation/SavePostValidationEvent.php:
```diff
@@ -49,7 +49,26 @@ final class SavePostValidationEvent extends SingleScheduledBackgroundTask implem
+		 * @param bool $enabled Enabled.
+		 * @internal
+		 */
+		return apply_filters( 'amp_temp_validation_cron_tasks_enabled', false );
```
I'm thinking in a subsequent PR to just remove the `SavePostValidationEvent` entirely. We'll have a weekly cron to re-check the most recently-published post aside from the user-initiated site scan in #6610. If a user has dev tools turned off, this will do a periodic health check without the overhead of doing validation for every single post when that data may not ever be used. What's more is that there is currently no garbage collection of validated URLs (although there needs to be), and without this the database would just keep silently getting larger and larger with validation data.

cc @delawski

dhaval-parekh commented:
Changes look good to me.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into dev-tool/5750-refactor-url-validation-cron

* 'develop' of github.com:ampproject/amp-wp:
  Reset ini config at tearDown
  Use temp filename for error log and reset ini before assertion
  Add test case for SupportData()
  Cover missing last line of error log
  Fix spacing issue
  Optimize function that reads error log.
  Remove unused variable
  Add test for amp-story-captions
  Update tests to fix failures
  Update amphtml spec to 530a457

westonruter commented:
Not sure why E2E would be failing here.

westonruter commented:
The E2E error:

>  TimeoutError: Element .done__links-container a:not([class*="--active"]) not found

This element is missing in the test:

![image](https://user-images.githubusercontent.com/134745/138148271-ed0c14ea-e8d4-46da-887f-e7e67f237d87.png)

westonruter committed:
Create a post before going to the onboarding wizard so there is something to preview

westonruter commented:
I haven't gotten the tests to run locally myself, so I'm shooting in the dark a bit. I suspect the issue is that now that `\AmpProject\AmpWP\Validation\ScannableURLProvider::get_urls()` only returns URLs on which AMP is enabled given the current template mode, if legacy Reader mode was initially selected then only singular posts will be returned among the URLs. 

@delawski So this is something else that needs to be added to #6650. In addition to re-fetching the scannable URLs for site scanning after changing the template mode, the Done screen of the Onboarding Wizard will also need to fetch the preview URLs _after_ the settings have been saved instead of exporting them up-front here:

https://github.com/ampproject/amp-wp/blob/ed75d8b1d05cd5f33946bbe4d2af4163f65a0d8d/src/Admin/OnboardingWizardSubmenuPage.php#L252

westonruter commented:
Alas, my supposed fix in fce116d didn't work. @delawski I believe you'd be able to quickly identify what is going on.

delawski committed:
Create test post and page using REST API

delawski committed:
Do not show site preview if there are no preview URLs

delawski committed:
Clean up site preview E2E test

delawski committed:
Fix AMP Settings Review panel E2E test

There needs to be at least one post or page for the "Browse Site" button to redirect to an AMP document. We're creating a test post before all tests and clean up the environment afterwards.

delawski commented:
> @delawski So this is something else that needs to be added to #6650. In addition to re-fetching the scannable URLs for site scanning after changing the template mode, the Done screen of the Onboarding Wizard will also need to fetch the preview URLs after the settings have been saved instead of exporting them up-front here

This has been already addressed in #6610 (https://github.com/ampproject/amp-wp/pull/6610/commits/ffe1c0e8fdbc2f1dd1106295d01daab4dff1ceac). We'll just have to combine changes done there with the update (https://github.com/ampproject/amp-wp/pull/6520/commits/43899d394458d3a9315350b60dbb306b7b091172) I pushed to this PR. What is does is it makes sure that no preview-related elements are rendered if there are no preview URLs (e.g. when there are no posts or pages).

> Alas, my supposed fix in fce116d didn't work. @delawski I believe you'd be able to quickly identify what is going on.

Your intuition here was correct. However, I just learnt that [`createNewPost`](https://github.com/WordPress/gutenberg/blob/e1db83135e5f9d6cbfb39d1aea1a5ca416d8773b/packages/e2e-test-utils/src/create-new-post.js) doesn't actually publish a post - it just opens up an editor and set some defaults.

We create and publish posts using REST API and that's the approach I took in https://github.com/ampproject/amp-wp/pull/6520/commits/8f2985a26b97251b211f2f41095b4160823b0fcf.

Note that besides creating a test `post`, we also create a test `page`. It's because in the Reader mode only singular templates are listed in the nav menu. It means that it would contain only a single element that would be active by default. It would lead to the same E2E error:

```
TimeoutError: Element .done__links-container a:not([class*="--active"]) not found
```

The changes I've introduced above (trashing all the post _and_ pages, in particular) resulted in side-effects in the AMP Settings Review panel. They've been fixed similarly to the ones in the Onboarding Wizard (https://github.com/ampproject/amp-wp/pull/6520/commits/8376583aed9670c6380677fd78c99c688153c7cd).

westonruter commented:
Thank you!

westonruter commented on assets/src/settings-page/site-review.js:
```diff
@@ -40,7 +40,7 @@ export function SiteReview() {
 	}
 
-	const previewPermalink = STANDARD === themeSupport ? HOME_URL : pairedUrlExamples[ pairedUrlStructure ][ 0 ];
+	const previewPermalink = STANDARD === themeSupport ? HOME_URL : pairedUrlExamples?.[ pairedUrlStructure ]?.[ 0 ] ?? HOME_URL;
```
This can be addressed in another PR, but even when in Standard mode we shouldn't assume that the `HOME_URL` has AMP enabled for it. The user could turn off the homepage template.

delawski replied:
Actually, the way we get the preview permalink changes in #6610 quite a bit. I guess this will likely need to be changed in context of your comment.

https://github.com/ampproject/amp-wp/blob/13fd8272144d5ec1437fc2e81bf4762b18079fcf/assets/src/components/site-scan-context-provider/index.js#L255-L259

delawski replied:
In fact, since `scannableUrls` that we're now getting are context-aware, I think we won't have to guess which page to use for a preview. We could just grab the first one from the list, correct?

westonruter replied:
That's right!

westonruter commented on tests/e2e/specs/admin/amp-options.js:
```diff
@@ -137,6 +137,30 @@ describe( 'Saving', () => {
+	afterAll( async () => {
+		await visitAdminPage( 'admin.php', 'page=amp-options' );
+
+		if ( testPost.id ) {
```
Minor thing: shouldn't this be:

```suggestion
		if ( testPost ) {
```

If `testPost` was not set, then this would be an error due to an undefined index?

delawski replied:
That's a good catch. I think we could actually go a step further like:

```suggestion
		if ( testPost?.id ) {
```

westonruter replied:
Yes, good.

delawski commented on tests/e2e/specs/amp-onboarding/done.js:
```diff
@@ -29,6 +34,43 @@ async function testCommonDoneStepElements() {
+	afterAll( async () => {
+		await visitAdminPage( 'admin.php', 'page=amp-options' );
+
+		if ( testPost.id ) {
```
```suggestion
		if ( testPost?.id ) {
```

delawski commented on tests/e2e/specs/amp-onboarding/done.js:
```diff
@@ -29,6 +34,43 @@ async function testCommonDoneStepElements() {
+				data: { force: true },
+			} ), testPost.id );
+		}
+		if ( testPage.id ) {
```
```suggestion
		if ( testPage?.id ) {
```

westonruter committed:
Harden check for post being set

Co-authored-by: Piotr Delawski <delawski@users.noreply.github.com>
